axum-extra/src/middleware/opentelemetry.rs

```diff
@@ -0,0 +1,513 @@
+        let http_target = uri
+            .path_and_query()
+            .map(|path_and_query| path_and_query.to_string())
+            .unwrap_or_else(|| uri.path().to_owned());
+
+        let client_ip = req
+            .extensions()
+            .get::<ConnectInfo<SocketAddr>>()
+            .map(|ConnectInfo(client_ip)| Cow::from(client_ip.to_string()))
+            .unwrap_or_default();
+
+        let request_id = req
+            .extensions()
+            .get::<RequestId>()
+            .and_then(|id| id.header_value().to_str().ok())
+            .unwrap_or_default();
+
+        let remote_context = extract_remote_context(req.headers());
+        let remote_span = remote_context.span();
+        let span_context = remote_span.span_context();
+        let trace_id = span_context
+            .is_valid()
+            .then(|| Cow::from(span_context.trace_id().to_string()))
+            .unwrap_or_default();
+
+        let span = tracing::info_span!(
+            "HTTP request",
+            http.client_ip = %client_ip,
+            http.flavor = %http_flavor(req.version()),
+            http.host = %host,
+            http.method = %http_method(req.method()),
+            http.route = %http_route,
+            http.scheme = %scheme,
+            http.status_code = Empty,
+            http.target = %http_target,
+            http.user_agent = %user_agent,
+            otel.kind = "server",
+            otel.status_code = Empty,
+            request_id = request_id,
+            trace_id = %trace_id,
+        );
+
+        tracing_opentelemetry::OpenTelemetrySpanExt::set_parent(&span, remote_context);
+
+        span
+    }
+}
+
+fn http_method(method: &Method) -> Cow<'static, str> {
+    match method {
+        &Method::CONNECT => "CONNECT".into(),
+        &Method::DELETE => "DELETE".into(),
+        &Method::GET => "GET".into(),
+        &Method::HEAD => "HEAD".into(),
+        &Method::OPTIONS => "OPTIONS".into(),
+        &Method::PATCH => "PATCH".into(),
+        &Method::POST => "POST".into(),
+        &Method::PUT => "PUT".into(),
+        &Method::TRACE => "TRACE".into(),
+        other => other.to_string().into(),
+    }
+}
+
+fn http_flavor(version: Version) -> Cow<'static, str> {
+    match version {
+        Version::HTTP_09 => "0.9".into(),
+        Version::HTTP_10 => "1.0".into(),
+        Version::HTTP_11 => "1.1".into(),
+        Version::HTTP_2 => "2.0".into(),
+        Version::HTTP_3 => "3.0".into(),
+        other => format!("{:?}", other).into(),
+    }
+}
+
+fn http_scheme(scheme: &Scheme) -> Cow<'static, str> {
+    if scheme == &Scheme::HTTP {
+        "http".into()
+    } else if scheme == &Scheme::HTTPS {
+        "https".into()
+    } else {
+        scheme.to_string().into()
+    }
+}
+
+// If remote request has no span data the propagator defaults to an unsampled context
+fn extract_remote_context(headers: &http::HeaderMap) -> opentelemetry::Context {
+    struct HeaderExtractor<'a>(&'a http::HeaderMap);
```

Comment:
Can we reuse existing from opentelemetry-http?

https://github.com/open-telemetry/opentelemetry-rust/blob/f28265947fad5de3027de98d08d79f197d49a733/opentelemetry-http/src/lib.rs#L24